ec2-spot-sagemaker-managed-spot-training/sagemaker-built-in-object-detection/tools/birdsOnEdge.py

```diff
@@ -0,0 +1,225 @@
+            os.mkfifo(result_path)
+        # This call will block until a consumer is available
+        with open(result_path, "w") as fifo_file:
+            while not self.stop_request.isSet():
+                try:
+                    # Write the data to the FIFO file. This call will block
+                    # meaning the code will come to a halt here until a consumer
+                    # is available.
+                    fifo_file.write(self.frame.tobytes())
+                except IOError:
+                    continue
+
+    def set_frame_data(self, frame):
+        """Method updates the image data. This currently encodes the
+        numpy array to jpg but can be modified to support other encodings.
+        frame - Numpy array containing the image data of the next frame
+                in the project stream.
+        """
+        ret, jpeg = cv2.imencode(".jpg", cv2.resize(frame, self.resolution))
+        if not ret:
+            raise Exception("Failed to set frame data")
+        self.frame = jpeg
+
+    def join(self):
+        self.stop_request.set()
+
+
+def greengrass_infinite_infer_run():
+    """Entry point of the lambda function"""
+
+    client.publish(topic=iot_topic, payload="Start of run loop...")
+
+    try:
+        # This object detection model is implemented as single shot detector (ssd), since
+        # the number of labels is small we create a dictionary that will help us convert
+        # the machine labels to human readable labels.
+        model_type = "ssd"
+
+        output_map = {}
+        with open("classes.txt") as f:
+            for line in f:
+                (key, val) = line.split()
+                output_map[int(key)] = val
+
+        client.publish(topic=iot_topic, payload="Classes to be detected: " + str(output_map))
+
+        # Create a local display instance that will dump the image bytes to a FIFO
+        # file that the image can be rendered locally.
+        local_display = LocalDisplay("480p")
+        local_display.start()
+
+        # The height and width of the training set images
+        input_height = 512
+        input_width = 512
+
+        # Load the model onto the GPU.
+
+        # optimize the model
+        client.publish(topic=iot_topic, payload="Optimizing model...")
+        ret, model_path = mo.optimize("deploy_ssd_resnet50_512", input_width, input_height)
+
+        # load the model
+        client.publish(topic=iot_topic, payload="Loading model...")
+        model = awscam.Model(model_path, {"GPU": 1})
+
+        client.publish(topic=iot_topic, payload="Custom object detection model loaded")
+
+        # Set the threshold for detection
+        detection_threshold = 0.40
+
+        # Do inference until the lambda is killed.
+        while True:
+            # Get a frame from the video stream
+            ret, frame = awscam.getLastFrame()
+            if not ret:
+                raise Exception("Failed to get frame from the stream")
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Using `Exception` and `BaseException` can make your code prone to errors and difficult to maintain. Instead, we recommend using one of the [Built-in Exceptions](https://docs.python.org/3/library/exceptions.html) or creating a custom exception class that is derived from `Exception` or one of its subclasses.